src/System.Management.Automation/engine/remoting/fanin/OutOfProcTransportManager.cs

```diff
@@ -1172,6 +1154,80 @@ internal override void CreateAsync()
             SendOneItem();
         }
 
+        private void StartRedirectionReaderThreads(Process serverProcess)
+        {
+            Thread outputThread = new Thread(ProcessOutputData);
+            outputThread.IsBackground = true;
+            outputThread.Name = "Out-of-Proc Job Output Thread";
+
+            Thread errorThread = new Thread(ProcessErrorData);
+            errorThread.IsBackground = true;
+            errorThread.Name = "Out-of-Proc Job Error Thread";
+
+            outputThread.Start(serverProcess.StandardOutput);
+            errorThread.Start(serverProcess.StandardError);
+        }
+
+        private void ProcessOutputData(object arg)
+        {
+            if (arg is StreamReader reader)
+            {
+                try
+                {
+                    string data = reader.ReadLine();
+                    while (data != null)
+                    {
+                        HandleOutputDataReceived(data);
```

Comment:
Since it runs in separate thread no race condition possible?

Reply:
Race condition with what?

Reply:
Previously the handler works in one thread for all jobs. Now it is called from different threads. It makes me think about some synchronization

Reply:
> Previously the handler works in one thread for all jobs

That's not true. The handler was registered for each job child powershell process, and now it's the same. The different is only previously it was using a background thread, now it's using a dedicated thread.